### ISSUE TITLE
rpc server: Subscription buffer limit exceeded

### ISSUE THREAD

**Issue by stakeworld** [paritytech/polkadot-sdk#4184]
I'm not sure if it is a bug or intended behaviour; sometimes i'm getting:
 `Subscription buffer limit=16 exceeded for subscription=state_storage conn_id=(...); dropping subscription` 
on my public rpc server nodes. I cannot find were the limit=16 comes from. I can find some limits in:

https://github.com/paritytech/jsonrpsee/blob/master/client/ws-client/src/lib.rs

> max_concurrent_requests: 256,
> max_buffer_capacity_per_subscription: 1024,

or 
https://github.com/paritytech/jsonrpsee/blob/master/client/http-client/src/client.rs

> max_concurrent_requests: 256,

were I notice that the http-client version does not have a  `max_buffer_capacity_per_subscription` element. And there is not a limit of 16 somewere.

Anyone knows if i'm missing a setting somewere or if the error is a problem at all?

**Comment by stakeworld** [comment]
@niklasad1, thanks for the explanation! It happens around 1-2 times per day on different nodes (they are load balanced polkadot public rpc endpoints). Resources (memory/network etc) are sufficient. It could of course be a slow client. I will try to increase the buffer-capacity for a test to see if it helps.

**Comment by stakeworld** [comment]
@niklasad1, I put `--rpc-message-buffer-capacity-per-connection 32` in the startup options but i'm still getting `Subscription buffer limit=16 exceeded for subscription=chain_newHead conn_id=(...); dropping subscription` errors... Do you have any pointers?

**Comment by niklasad1** [comment]
You need to increase to something bigger --rpc-message-buffer-capacity-per-connection 64  it's the default.

I suggest to start with 128 :)

**Comment by stakeworld** [comment]
> You need to increase to something bigger because `--rpc-message-buffer-capacity-per-connection 64` is the default.

I got an increase in errors so that makes sense ;)

Increasing to 128.

**Comment by stakeworld** [comment]
@niklasad1 : With 128 I haven't seen the error coming back so it seems a more stable value, at least for a busy rpc node, thanks for the pointers.

**Comment by stakeworld** [comment]
Hi @niklasad1, i've been monitoring for a while now and even bumped the buffer capacity to 512 (`--rpc-message-buffer-capacity-per-connection 512`) but the buffer limit errors keep coming (and the buffer limit in the error stays on 16). Do you maybe have some insights where to look?

> 2024-05-17 07:30:06 Subscription buffer limit=16 exceeded for subscription=state_storage conn_id=905687; dropping subscription

**Comment by niklasad1** [comment]
Yes, it's still hardcoded to max 16 per subscription but you are increasing the global message buffer on the server with `--rpc-message-buffer-per-connection` so it will allow more buffered messages.

Lemme write some prometheus metrics for this so we can monitor it better and perhaps the hardcoded limit is simply wrong.

**Comment by stakeworld** [comment]
> Lemme write some prometheus metrics for this so we can monitor it better and perhaps the hardcoded limit is simply wrong.
I could be its fine for a normal node but maybe an active rpc node needs something higher?

> I still think it's just client's that are too slow....
If the conclusion is that it's just a slow client problem and doesn't have a negative influence on rpc functioning than I will just filter them out.

Let me know if I can help with providing some prometheus or log metrics from an active node.
